scheduler/preemption.go

```diff
@@ -0,0 +1,572 @@
+
+		// Only include if the alloc has a network device
+		if len(networks) > 0 {
+			device := networks[0].Device
+			allocsForDevice := deviceToAllocs[device]
+			allocsForDevice = append(allocsForDevice, alloc)
+			deviceToAllocs[device] = allocsForDevice
+		}
+	}
+
+	// If no existing allocations use network resources, return early
+	if len(deviceToAllocs) == 0 {
+		return nil
+	}
+
+	var allocsToPreempt []*structs.Allocation
+
+	met := false
+	freeBandwidth := 0
+
+	for device, currentAllocs := range deviceToAllocs {
+		totalBandwidth := netIdx.AvailBandwidth[device]
+
+		// If the device doesn't have enough total available bandwidth, skip
+		if totalBandwidth < MbitsNeeded {
+			continue
+		}
+
+		// Track how much existing free bandwidth we have before preemption
+		freeBandwidth = totalBandwidth - netIdx.UsedBandwidth[device]
+
+		preemptedBandwidth := 0
+
+		// Reset allocsToPreempt since we don't want to preempt across devices for the same task
+		allocsToPreempt = nil
+
+		// Build map from used reserved ports to allocation
+		usedPortToAlloc := make(map[int]*structs.Allocation)
+
+		// First try to satisfy needed reserved ports
+		if len(reservedPortsNeeded) > 0 {
+			for _, alloc := range currentAllocs {
+				for _, tr := range alloc.TaskResources {
+					reservedPorts := tr.Networks[0].ReservedPorts
+					for _, p := range reservedPorts {
+						usedPortToAlloc[p.Value] = alloc
+					}
+				}
+			}
+
+			// Look for allocs that are using reserved ports needed
+			for _, port := range reservedPortsNeeded {
+				alloc, ok := usedPortToAlloc[port.Value]
+				if ok {
+					preemptedBandwidth += alloc.Resources.Networks[0].MBits
+					allocsToPreempt = append(allocsToPreempt, alloc)
+				}
+			}
+
+			// Remove allocs that were preempted to satisfy reserved ports
+			currentAllocs = structs.RemoveAllocs(currentAllocs, allocsToPreempt)
+		}
+
+		// If bandwidth requirements have been met, stop
+		if preemptedBandwidth+freeBandwidth >= MbitsNeeded {
+			break
+		}
+
+		// Split by priority
+		allocsByPriority := filterAndGroupPreemptibleAllocs(p.jobPriority, currentAllocs)
+
+		for _, allocsGrp := range allocsByPriority {
+			allocs := allocsGrp.allocs
+
+			// Sort by distance function
+			sort.Slice(allocs, func(i, j int) bool {
+				return p.distanceComparatorForNetwork(allocs, networkResourceAsk, i, j)
+			})
+
+			// Iterate over allocs until end of if requirements have been met
+			for _, alloc := range allocs {
+				preemptedBandwidth += alloc.Resources.Networks[0].MBits
+				allocsToPreempt = append(allocsToPreempt, alloc)
+				if preemptedBandwidth+freeBandwidth >= MbitsNeeded {
+					met = true
+					break
+				}
+			}
+
+			// If we met bandwidth needs we can break out of iterating by priority within a device
+			if met {
+				break
+			}
+		}
+		// If we met bandwidth needs we don't need to examine the next network device
+		if met {
+			break
+		}
+	}
+	if len(allocsToPreempt) == 0 {
```

Comment:
This isn't correct. If you go through all devices and the final device you add some allocs to allocsToPreempt but never meet the required mbits this slice won't be empty but you haven't met the resource request